File: src/test/kotlin/info/davidvedvick/seis739/catpics/users/given an existing user.kt
package info.davidvedvick.seis739.catpics.users

import info.davidvedvick.seis739.catpics.security.AuthenticatedCatEmployee
import info.davidvedvick.seis739.catpics.security.CatEmployeeEntry
import info.davidvedvick.seis739.catpics.security.UnauthenticatedCatEmployee
import io.mockk.coEvery
import io.mockk.every
import io.mockk.mockk
import kotlinx.coroutines.runBlocking
import org.amshove.kluent.`should be equal to`
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class `given an existing user` {
    @Nested
    inner class `when logging the user in` {
        private val services by lazy {
            CatEmployeeEntry(
                mockk {
                    coEvery { findByEmail("2l9L") } returns CatEmployee(822, "2l9L", "MobWbSRg", true)
                },
                mockk {
                    every { encode("zc89") } returns "MobWbSRg"
                    every { matches("zc89", "MobWbSRg") } returns true
                },
            )
        }

        private var user: AuthenticatedCatEmployee? = null

        @BeforeAll
        fun act() {
            runBlocking {
                user = services.authenticate(UnauthenticatedCatEmployee("2l9L", "zc89")) as? AuthenticatedCatEmployee
            }
        }

        @Test
        fun `then the email is correct`() {
            user?.email `should be equal to` "2l9L"
        }

        @Test
        fun `then the password is correct`() {
            user?.password `should be equal to` "zc89"
        }
    }
}

File: server/users/specs/given an existing user.ts
import {beforeAll, describe, expect, test} from "@jest/globals";
import {AuthenticatedCatEmployee} from "../AuthenticatedCatEmployee.js";
import CatEmployeeEntry from "../CatEmployeeEntry.js";
import CatEmployee from "../CatEmployee.js";
import {ManageCatEmployees} from "../ManageCatEmployees.js";
import {UnauthenticatedCatEmployee} from "../UnauthenticatedCatEmployee.js";
import Encoder from "../../security/Encoder.js";

describe("given an existing user", () => {
    describe("when logging the user in", () => {
        let user: AuthenticatedCatEmployee;

        beforeAll(async () => {
            const catEmployeeEntry = new CatEmployeeEntry({
                findByEmail(email: string): Promise<CatEmployee | null> {
                    return Promise.resolve(email == "2l9L" ? {
                        email: "2l9L",
                        password: "MobWbSRg",
                        isEnabled: true,
                        id: 822,
                    } : null);
                }
            } as ManageCatEmployees, {
                matches(rawPassword: string, encodedPassword: string): boolean {
                    return rawPassword == "zc89" && encodedPassword == "MobWbSRg";
                }
            } as Encoder);

            user = await catEmployeeEntry.authenticate({
                email: "2l9L",
                password: "zc89",
            });
        });

        test("then the email is correct", () => {
            expect(user.email).toBe("2l9L");
        });

        test("then the password is correct", () => {
            expect(user.password).toBe("MobWbSRg");
        });

        test("then the user is authenticated", () => {
            expect(user).toBeInstanceOf(AuthenticatedCatEmployee);
        });
    });

    describe("and the password does not match", () => {
        let user: AuthenticatedCatEmployee;

        beforeAll(async () => {
            const catEmployeeEntry = new CatEmployeeEntry({
                findByEmail(email: string): Promise<CatEmployee | null> {
                    return Promise.resolve(email == "2l9L" ? {
                        email: "2l9L",
                        password: "69FM2Vw",
                        isEnabled: true,
                        id: 584,
                    } : null);
                }
            } as ManageCatEmployees, {
                matches(rawPassword: string, encodedPassword: string): boolean {
                    return rawPassword == "zc89" && encodedPassword == "MobWbSRg";
                }
            } as Encoder);

            user = await catEmployeeEntry.authenticate({
                email: "2l9L",
                password: "zc89",
            });
        });

        test("then the email is correct", () => {
            expect(user.email).toBe("2l9L");
        });

        test("then the password is correct", () => {
            expect(user.password).toBe("69FM2Vw");
        });

        test("then the user is not authenticated", () => {
            expect(user).toBeInstanceOf(UnauthenticatedCatEmployee);
        });
    });
});